core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -664,41 +646,40 @@ class KafkaController(val config: KafkaConfig,
-    reassignedPartitionContext.ongoingReassignmentOpt match {
-      case Some(ongoingAssignment) =>
-        val originalAssignment = ongoingAssignment.previousAssignment
-        assert(ongoingAssignment.isBeingReassigned)
-        assert(!originalAssignment.isBeingReassigned)
+   * Update the current assignment state in Zookeeper and in memory. If a reassignment is already in
+   * progress, then the new reassignment will supplant it and some replicas will be shutdown.
+   *
+   * Note that due to the way we compute the original replica set, we have no guarantee that a revert
+   * would put it in the same order.
+   *
+   * @param topicPartition The reassigning partition
+   * @param reassignment The new reassignment context
+   *
+   * @return The updated assignment state
+   */
+  private def updateCurrentReassignment(topicPartition: TopicPartition,
+                                        reassignment: ReassignmentContext): ReplicaAssignment = {
+    val currentAssignment = controllerContext.partitionFullReplicaAssignment(topicPartition)
+    val updatedAssignment = currentAssignment.reassignTo(reassignment.newReplicas)
 
-        val unnecessaryReplicas = ongoingAssignment.replicas.filterNot(originalAssignment.replicas.contains(_))
-        val (overlappingReplicas, replicasToRemove) = unnecessaryReplicas.partition(reassignedPartitionContext.newReplicas.contains(_))
-        // RS = ORS + OVRS, AR = OVRS, RR = []
-        val intermediateReplicaAssignment = PartitionReplicaAssignment(originalAssignment.replicas ++ overlappingReplicas, overlappingReplicas, Seq())
+    if (currentAssignment != updatedAssignment) {
+      info(s"Updated assignment of partition $topicPartition from $currentAssignment to $updatedAssignment")
 
-        if (isDebugEnabled)
-          debug(s"Reverting previous reassignment $originalAssignment (we were in the " +
-            s"process of an ongoing reassignment with target replicas ${ongoingAssignment.targetReplicas.mkString(",")} (" +
-            s"Overlapping replicas: ${overlappingReplicas.mkString(",")}, Replicas to remove: ${replicasToRemove.mkString(",")})")
+      // A1 and A2: Update zk and cached state
+      updateReplicaAssignmentForPartition(topicPartition, updatedAssignment)
+      controllerContext.updatePartitionFullReplicaAssignment(topicPartition, updatedAssignment)
 
-        // Set RS = ORS + OVRS, AR = OVRS, RR = [] in memory.
-        controllerContext.updatePartitionFullReplicaAssignment(topicPartition, intermediateReplicaAssignment)
+      // If there is a reassignment already in progress, then some of the currently adding replicas
+      // may be eligible for immediate removal, in which case we need to stop the replicas.
+      val unneededReplicas = currentAssignment.replicas.diff(updatedAssignment.replicas)
+      if (unneededReplicas.nonEmpty)
+        stopRemovedReplicasOfReassignedPartition(topicPartition, unneededReplicas)
```

Comment:
I was thinking it was unnecessary because we already get an epoch bump in A3, but in the case of a cancellation, we may not hit that case at all. So perhaps we still need it. It's a bit tough at the moment to see all of the epoch bumps because some of them are hidden beneath replica state changes. Let me look at this more carefully.

Reply:
Ok, I think I convinced myself that it is unnecessary. Whether we go down the phase A path or phase B path, we will get an epoch bump. In phase B, the epoch will be bumped in `moveReassignedPartitionLeaderIfRequired`.  Does that seem right?

Reply:
I also traced it down - I think you're right.